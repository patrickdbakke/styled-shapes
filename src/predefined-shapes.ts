export type Point = { x: number; y: number };
export const polygon = (n: number, offsetAngle = 0): string => {
    let str = 'M';
    for (let i = 0; i < n; i++) {
        const angle = offsetAngle + Math.PI + (-1 * 2.0 * Math.PI * i) / n;
        const x = 0.5 * Math.sin(angle) + 0.5;
        const y = 0.5 * Math.cos(angle) + 0.5;
        str += ` ${x},${y}`;
    }
    str += ' Z';
    return str;
};

export const triangle = polygon(3);
export const square = `M 0,0 1,0, 1,1, 0,1 Z`;
export const diamond = `M .5,0 1,.5, .5,1, 0,.5 Z`;
export const pentagon = polygon(5);
export const hexagon = polygon(6, Math.PI / 6);
export const heptagon = polygon(7);
export const octagon = polygon(8, Math.PI / 8);
export const nonagon = polygon(9);
export const decagon = polygon(12);
export const circle = polygon(100);
export const trapezoid = `M0.25 0.2113248654051872L0.75 0.2113248654051872 1 0.7886751345948129 0 0.7886751345948129Z`;

// prettier-ignore
export const plus = 'M 31.8532,29.8741 31.8532,12.521 42.1488,12.521 42.1488,29.8805 42.1967,29.8806 46.2484,29.8832 59.6052,29.8832 59.6052,40.1789 46.2562,40.1789 42.1493,40.1762 42.1488,40.206 42.1488,58.8513 31.8532,58.8513 31.8532,40.1694 31.7534,40.1693 27.9357,40.1667 14,40.1667 14,29.8711 27.9331,29.8711 31.7687,29.8737 Z';
// prettier-ignore
export const heart = 'M35.885 11.833c0-5.45-4.418-9.868-9.867-9.868-3.308 0-6.227 1.633-8.018 4.129-1.791-2.496-4.71-4.129-8.017-4.129-5.45 0-9.868 4.417-9.868 9.868 0 .772.098 1.52.266 2.241C1.751 22.587 11.216 31.568 18 34.034c6.783-2.466 16.249-11.447 17.617-19.959.17-.721.268-1.469.268-2.242z';
// prettier-ignore
export const moon = 'M313.283,275.341c-17.198,5.692-35.086,8.574-53.197,8.574c-93.377,0-169.347-75.96-169.347-169.353 c0-37.692,12.157-73.397,35.155-103.234c1.057-1.381,1.087-3.273,0.09-4.69c-1-1.398-2.795-2.011-4.443-1.453 C48.834,29.223,0,96.842,0,173.428C0,271.144,79.498,350.66,177.217,350.66c55.374,0,106.558-25.221,140.413-69.176 c1.075-1.375,1.105-3.267,0.108-4.672C316.735,275.395,314.927,274.789,313.283,275.341z';
// prettier-ignore
export const star = 'm25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z';
// prettier-ignore
export const egg = 'M3170 6659 c-585 -60 -1143 -473 -1582 -1172 -100 -160 -274 -504 -341 -677 -395 -1013 -421 -2054 -70 -2860 217 -501 537 -871 975 -1131 239 -142 477 -229 767 -280 145 -26 451 -36 604 -20 917 93 1661 711 1976 1639 70 204 117 420 146 667 23 189 23 566 1 778 -55 522 -198 1032 -419 1491 -190 394 -392 689 -656 960 -363 371 -735 566 -1158 606 -119 11 -122 11 -243 -1z';
// prettier-ignore
export const rainbow = 'M 0 .75 A .5 .75 0 0 1 1 .75 L .75 .75 A .2 .3 0 1 0 .25 .75';
