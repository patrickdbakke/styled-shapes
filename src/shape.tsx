import Flex, { FlexProps } from '@chakra-ui/core/dist/Flex';
import md5 from 'md5';
import React from 'react';
import svgpath from 'svgpath';
import { SVGBorder } from './utils/border';
import { WithShadowFilters } from './utils/shadow';
import { useRef } from './utils/use-ref';

const ns = 'http://www.w3.org/2000/svg';
const svgRoot = document.createElementNS(ns, 'svg');
svgRoot.setAttribute('id', 'shapes-path-root');
const defs = document.createElementNS(ns, 'defs');
svgRoot.appendChild(defs);
document.body.appendChild(svgRoot);

export type ShapeProps = FlexProps &
    React.DOMAttributes<HTMLDivElement> & {
        path: string;
        border?: string;
        borderLeft?: string;
        borderRight?: string;
        borderTop?: string;
        borderBottom?: string;
        boxShadow?: string;
        key?: string | number;
        style?: {
            [key: string]: unknown;
        };
        shapeProps?: {
            width?: string;
            height?: string;
            style?: ShapeProps['style'];
        };
    };

type ContentSize = {
    width?: string;
    height?: string;
    marginTop?: string;
    marginLeft?: string;
};
export type NamedShapeProps = Omit<ShapeProps, 'path'>;
export const Shape: React.FC<ShapeProps> = ({
    path = 'M 0,0,0,1,1,1,1,0 Z',
    children,
    color = 'inherit',
    background = 'inherit',
    backgroundImage,
    backgroundSize,
    backgroundColor,
    backgroundPosition,
    backgroundAttachment,
    backgroundRepeat,
    boxShadow,
    width: w,
    height: h,
    top,
    left,
    right,
    bottom,
    border,
    borderTop = border,
    borderRight = border,
    borderBottom = border,
    borderLeft = border,
    padding,
    paddingTop = padding,
    paddingRight = padding,
    paddingBottom = padding,
    paddingLeft = padding,
    margin,
    marginTop = margin,
    marginRight = margin,
    marginBottom = margin,
    marginLeft = margin,
    overflow = 'visible',
    display,
    flexDirection,
    flex,
    flexBasis,
    shapeProps,
    ...props
}: ShapeProps) => {
    const pathId = `path-${md5(path)}`;

    const unscaledPath: typeof svgpath = svgpath(path)
        .unshort()
        .unarc()
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        .abs();
    let minX = Infinity;
    let maxX = -Infinity;
    let minY = Infinity;
    let maxY = -Infinity;
    unscaledPath.iterate((_segment, _i, x, y) => {
        minX = Math.min(minX, x);
        maxX = Math.max(maxX, x);
        minY = Math.min(minY, y);
        maxY = Math.max(maxY, y);
    });
    const spanX = maxX - minX;
    const spanY = maxY - minY;
    const maxDist = Math.max(spanX, spanY);
    const svgPath = unscaledPath.scale(1 / maxDist);
    let width = w;
    let height = h;
    if (typeof width === 'undefined' && typeof height === 'undefined') {
        width = '1em';
        height = '1em';
    } else if (typeof width !== 'undefined' && typeof height === 'undefined') {
        height = `calc(${width} * ${spanY / spanX})`;
    } else if (typeof height !== 'undefined' && typeof width === 'undefined') {
        width = `calc(${height} * ${spanX / spanY})`;
    }
    const [contentSize, setContentSize] = React.useState<ContentSize>({
        width,
        height,
    } as ContentSize);
    const ref = useRef<HTMLElement>(null);
    React.useEffect(() => {
        if (ref.current) {
            const style = window.getComputedStyle(ref.current);
            const currentWidth = parseFloat(style.width);
            const currentHeight = parseFloat(style.height);
            if (currentWidth / currentHeight > spanX / spanY) {
                const newWidth = (currentHeight * spanX) / spanY;
                setContentSize({
                    height: contentSize.height,
                    width: `${newWidth}px`,
                });
            } else if (currentWidth / currentHeight < spanX / spanY) {
                const newHeight = (currentWidth * spanY) / spanX;
                setContentSize({
                    height: `${newHeight}px`,
                    width: contentSize.width,
                });
            }
        }
    }, [ref, ref.current]);
    const bg = {
        background,
        backgroundImage,
        backgroundSize,
        backgroundColor,
        backgroundPosition,
        backgroundAttachment,
        backgroundRepeat,
    };
    const b = {
        border,
        borderTop,
        borderLeft,
        borderRight,
        borderBottom,
    };
    const p = {
        padding,
        paddingTop,
        paddingLeft,
        paddingRight,
        paddingBottom,
    };
    const m = {
        margin,
        marginTop,
        marginRight,
        marginBottom,
        marginLeft,
    };
    const parentProps = {
        position: 'relative' as 'relative',
        display: 'inline-block',
        width,
        height,
        top,
        left,
        right,
        bottom,
        ...m,
    };
    const wrapperProps = {
        background: 'transparent',
        justifyContent: 'center',
        alignItems: 'center',
        position: 'absolute' as 'absolute',
        top: '50%',
        left: '50%',
        transform: `translateX(-50%) translateY(-50%)`,
        color,
        zIndex: 3,
        margin: 'auto',
        opacity: props.opacity as number | undefined,
        ...props,
        style: {
            ...contentSize,
            clipPath: overflow !== 'visible' ? `url(#${pathId})` : undefined,
        },
    };
    const shapeprops = {
        zIndex: 2,
        ...bg,
        ...shapeProps,
        transform: `translateX(-50%) translateY(-50%)`,
        style: {
            ...contentSize,
            ...(shapeProps?.style || {}),
            clipPath: `url(#${pathId})`,
        },
    };
    const contentProps = {
        ...p,
        fontSize: '.5em',
        boxSizing: 'content-box' as 'content-box',
        justifyContent: 'center',
        alignItems: 'center',
        display,
        flexDirection,
        flex,
        flexBasis,
    };

    if (document.querySelectorAll(`#${pathId}`).length < 1) {
        const clipPath = document.createElementNS(ns, 'clipPath');
        clipPath.setAttribute('id', pathId);
        clipPath.setAttribute('clipPathUnits', 'objectBoundingBox');
        const pathElement = document.createElementNS(ns, 'path');
        pathElement.setAttribute('d', svgPath.toString());
        clipPath.appendChild(pathElement);
        defs.appendChild(clipPath);
    }

    return (
        <Flex ref={ref} {...parentProps}>
            <Flex position="relative" margin="auto" height="100%" width="100%">
                <Flex {...wrapperProps}>
                    <Flex {...contentProps}>{children}</Flex>
                </Flex>
                {contentSize.width && contentSize.height && (
                    <>
                        <WithShadowFilters shadow={boxShadow} {...contentSize}>
                            <Flex
                                position="absolute"
                                transformOrigin="50% 50%"
                                top="50%"
                                left="50%"
                                {...shapeprops}
                            />
                        </WithShadowFilters>
                        <SVGBorder path={svgPath} {...b} {...contentSize} />
                    </>
                )}
            </Flex>
        </Flex>
    );
};
