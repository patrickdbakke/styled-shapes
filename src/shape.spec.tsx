import { Flex } from '@chakra-ui/core';
import { render } from '@testing-library/react';
import * as React from 'react';
import { octagon } from './predefined-shapes';
import { Shape } from './shape';

const oneTick = () => new Promise((resolve) => setTimeout(resolve));
describe('the svg path', () => {
    describe('shape 1 (stop sign)', () => {
        beforeEach(async () => {
            render(
                <Shape
                    path={octagon}
                    background="#CF242F"
                    border="3px solid white"
                    fontFamily="sans-serif"
                    margin=".5em"
                    width="2em"
                    height="2em"
                    overflow="hidden"
                    color="white"
                    display="flex"
                    flexDirection="column"
                    data-testid="shape"
                    boxShadow="inset 0 0 10px 3px black, 0 0 3px 3px black"
                >
                    <Flex fontSize="14px">this is html</Flex>
                </Shape>,
            );
            await oneTick();
        });
        it('should match the snapshot', () => {
            expect(document.body).toMatchSnapshot();
        });
    });
    describe('shape 2 (hexagon)', () => {
        beforeEach(async () => {
            render(
                <Shape
                    path="M0.5 1L0.9330127018922193 0.75 0.9330127018922193 0.75 0.9330127018922193 0.25 0.5 0 0.0669872981077807 0.25 0.0669872981077807 0.75 0.5 1Z"
                    background="radial-gradient(#fff, #ddd, #999, #aaa)"
                    border="1px solid black"
                    boxShadow=".05em .05em .125em 3px rgba(0,0,0,.5), .125em .125em .5em rgba(0,0,0,.5)"
                    margin=".5em"
                />,
            );
            await oneTick();
        });
        it('should match the snapshot', () => {
            expect(document.body).toMatchSnapshot();
        });
    });
});
