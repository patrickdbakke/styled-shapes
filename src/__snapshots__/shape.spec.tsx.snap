// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the svg path shape 1 (stop sign) should match the snapshot 1`] = `
<body>
  <div
    class="Toaster"
    id="react-toast"
  >
    <span
      class="Toaster__manager-top"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; top: 0px; right: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-top-left"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; top: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-top-right"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; top: 0px; right: 0px;"
    />
    <span
      class="Toaster__manager-bottom-left"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-bottom"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; bottom: 0px; right: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-bottom-right"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; right: 0px;"
    />
  </div>
  <div>
    <div
      class="css-kqywxn"
    >
      <div
        class="css-1c4rusl"
      >
        <div
          class="css-8meu9y"
          data-testid="shape"
          style="width: 2em; height: 2em; clip-path: url(#path-f032a833e240e62937ad9533bc337a97);"
        >
          <div
            class="css-19wy3bq"
          >
            <div
              class="css-1rjzkmk"
            >
              this is html
            </div>
          </div>
        </div>
        <div
          class="css-3sakfm"
          style="filter: url(#filter-81618823da0884cccd5778b780210f0e);"
        >
          <div
            class="css-1ex0plv"
            style="clip-path: url(#path-f032a833e240e62937ad9533bc337a97);"
          />
        </div>
        <div
          class="css-s8qk35"
        >
          <svg
            fill="transparent"
            height="100%"
            style="position: absolute; top: 0px; left: 0px;"
            viewBox="0 0 1 1"
            width="100%"
          >
            <path
              d="M 0.29289321881345254 0 L 0.7071067811865476 6.245004513516506e-17 M 0.7071067811865476 6.245004513516506e-17 L 1.0000000000000002 0.29289321881345254"
              fill="transparent"
              stroke="white"
              stroke-width="NaNpx"
            />
            <path
              d="M 1.0000000000000002 0.29289321881345254 L 1.0000000000000002 0.7071067811865477"
              fill="transparent"
              stroke="white"
              stroke-width="NaNpx"
            />
            <path
              d="M 1.0000000000000002 0.7071067811865477 L 0.7071067811865476 1.0000000000000002 M 0.7071067811865476 1.0000000000000002 L 0.2928932188134524 1.0000000000000002"
              fill="transparent"
              stroke="white"
              stroke-width="NaNpx"
            />
            <path
              d="M 0.2928932188134524 1.0000000000000002 L 0 0.7071067811865477 M 0 0.7071067811865477 L 0 0.29289321881345254 M 0 0.29289321881345254 L 0.29289321881345254 0  "
              fill="transparent"
              stroke="white"
              stroke-width="NaNpx"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    id="once-d4cdfd84b312b43f1b8054427089f6dc"
  >
    <svg>
      <clippath
        clipPathUnits="objectBoundingBox"
        id="path-f032a833e240e62937ad9533bc337a97"
      >
        <path
          d="M0.29289321881345254 0L0.7071067811865476 6.245004513516506e-17 1.0000000000000002 0.29289321881345254 1.0000000000000002 0.7071067811865477 0.7071067811865476 1.0000000000000002 0.2928932188134524 1.0000000000000002 0 0.7071067811865477 0 0.29289321881345254Z"
        />
      </clippath>
    </svg>
  </div>
  <div
    id="once-fce492877a0be2c6885647e0089248d7"
  >
    <svg
      style="overflow: hidden; height: 0px; width: 0px; opacity: 0;"
    >
      <defs>
        <filter
          color-interpolation-filters="sRGB"
          height="300%"
          id="filter-81618823da0884cccd5778b780210f0e"
          width="300%"
          x="-100%"
          y="-100%"
        >
          <feflood
            flood-color="black"
            result="color-0"
          />
          <fecomposite
            in="color-0"
            in2="SourceGraphic"
            operator="out"
            result="inverse-0"
          />
          <femorphology
            in="filter-0"
            operator="dilate"
            radius="3"
            result="dilated-0"
          />
          <feoffset
            dx="0"
            dy="0"
            in="dilated-0"
            result="offsetted-0"
          />
          <fegaussianblur
            in="offsetted-0"
            result="blurred-0"
            stdDeviation="5"
          />
          <fecomposite
            in="blurred-0"
            in2="SourceGraphic"
            operator="in"
            result="shadow-0"
          />
          <fecomposite
            in="shadow-0"
            in2="SourceGraphic"
            operator="atop"
            result="filter-0"
          />
          <feflood
            flood-color="black"
            result="flooded-1"
          />
          <femorphology
            in="SourceAlpha"
            operator="dilate"
            radius="3"
            result="dilated-1"
          />
          <feoffset
            dx="0"
            dy="0"
            in="dilated-1"
            result="offset-1"
          />
          <fegaussianblur
            in="offset-1"
            result="blurred-1"
            stdDeviation="1.5"
          />
          <fecomposite
            in="flooded-1"
            in2="blurred-1"
            operator="in"
            result="shadow-1"
          />
          <femerge
            result="filter-1"
          >
            <femergenode
              in="shadow-1"
            />
            <femergenode
              in="SourceGraphic"
            />
          </femerge>
          <femerge>
            <femergenode
              in="filter-1"
            />
            <femergenode
              in="filter-0"
            />
          </femerge>
        </filter>
      </defs>
    </svg>
  </div>
</body>
`;

exports[`the svg path shape 2 (hexagon) should match the snapshot 1`] = `
<body>
  <div
    class="Toaster"
    id="react-toast"
  >
    <span
      class="Toaster__manager-top"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; top: 0px; right: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-top-left"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; top: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-top-right"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; top: 0px; right: 0px;"
    />
    <span
      class="Toaster__manager-bottom-left"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-bottom"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; bottom: 0px; right: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-bottom-right"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; right: 0px;"
    />
  </div>
  <div
    id="once-d4cdfd84b312b43f1b8054427089f6dc"
  >
    <svg>
      <clippath
        clipPathUnits="objectBoundingBox"
        id="path-f032a833e240e62937ad9533bc337a97"
      >
        <path
          d="M0.29289321881345254 0L0.7071067811865476 6.245004513516506e-17 1.0000000000000002 0.29289321881345254 1.0000000000000002 0.7071067811865477 0.7071067811865476 1.0000000000000002 0.2928932188134524 1.0000000000000002 0 0.7071067811865477 0 0.29289321881345254Z"
        />
      </clippath>
    </svg>
  </div>
  <div
    id="once-fce492877a0be2c6885647e0089248d7"
  >
    <svg
      style="overflow: hidden; height: 0px; width: 0px; opacity: 0;"
    >
      <defs>
        <filter
          color-interpolation-filters="sRGB"
          height="300%"
          id="filter-81618823da0884cccd5778b780210f0e"
          width="300%"
          x="-100%"
          y="-100%"
        >
          <feflood
            flood-color="black"
            result="color-0"
          />
          <fecomposite
            in="color-0"
            in2="SourceGraphic"
            operator="out"
            result="inverse-0"
          />
          <femorphology
            in="filter-0"
            operator="dilate"
            radius="3"
            result="dilated-0"
          />
          <feoffset
            dx="0"
            dy="0"
            in="dilated-0"
            result="offsetted-0"
          />
          <fegaussianblur
            in="offsetted-0"
            result="blurred-0"
            stdDeviation="5"
          />
          <fecomposite
            in="blurred-0"
            in2="SourceGraphic"
            operator="in"
            result="shadow-0"
          />
          <fecomposite
            in="shadow-0"
            in2="SourceGraphic"
            operator="atop"
            result="filter-0"
          />
          <feflood
            flood-color="black"
            result="flooded-1"
          />
          <femorphology
            in="SourceAlpha"
            operator="dilate"
            radius="3"
            result="dilated-1"
          />
          <feoffset
            dx="0"
            dy="0"
            in="dilated-1"
            result="offset-1"
          />
          <fegaussianblur
            in="offset-1"
            result="blurred-1"
            stdDeviation="1.5"
          />
          <fecomposite
            in="flooded-1"
            in2="blurred-1"
            operator="in"
            result="shadow-1"
          />
          <femerge
            result="filter-1"
          >
            <femergenode
              in="shadow-1"
            />
            <femergenode
              in="SourceGraphic"
            />
          </femerge>
          <femerge>
            <femergenode
              in="filter-1"
            />
            <femergenode
              in="filter-0"
            />
          </femerge>
        </filter>
      </defs>
    </svg>
  </div>
  <div>
    <div
      class="css-hmi9ye"
    >
      <div
        class="css-1c4rusl"
      >
        <div
          class="css-ay8fpe"
          style="width: 0.8660254037844386px; height: 0.8660254037844386px; clip-path: url(#path-43f2fd1f4be4cb8199b37dd28e4589bb);"
        >
          <div
            class="css-ddh4jq"
          />
        </div>
        <div
          class="css-1g4mdh1"
          style="filter: url(#filter-95f111bf6c90801da04c3e2220c21389);"
        >
          <div
            class="css-1ec1a0"
            style="clip-path: url(#path-43f2fd1f4be4cb8199b37dd28e4589bb);"
          />
        </div>
        <div
          class="css-106pygv"
        >
          <svg
            fill="transparent"
            height="100%"
            style="position: absolute; top: 0px; left: 0px;"
            viewBox="0 0 1 1"
            width="100%"
          >
            <path
              d="M 0.5 1 L 0.9330127018922193 0.75 M 0.0669872981077807 0.75 L 0.5 1"
              fill="transparent"
              stroke="black"
              stroke-width="NaNpx"
            />
            <path
              d="M 0.0669872981077807 0.25 L 0.0669872981077807 0.75"
              fill="transparent"
              stroke="black"
              stroke-width="NaNpx"
            />
            <path
              d="M 0.9330127018922193 0.75 L 0.9330127018922193 0.75 M 0.9330127018922193 0.25 L 0.5 0 M 0.5 0 L 0.0669872981077807 0.25 M 0.5 1 L 0.5 1  "
              fill="transparent"
              stroke="black"
              stroke-width="NaNpx"
            />
            <path
              d="M 0.9330127018922193 0.75 L 0.9330127018922193 0.25"
              fill="transparent"
              stroke="black"
              stroke-width="NaNpx"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    id="once-5dda08ac87d9f3a77446f01f32a62639"
  >
    <svg>
      <clippath
        clipPathUnits="objectBoundingBox"
        id="path-43f2fd1f4be4cb8199b37dd28e4589bb"
      >
        <path
          d="M0.5 1L0.9330127018922193 0.75 0.9330127018922193 0.75 0.9330127018922193 0.25 0.5 0 0.0669872981077807 0.25 0.0669872981077807 0.75 0.5 1Z"
        />
      </clippath>
    </svg>
  </div>
  <div
    id="once-2dc643fc620a5d7bb8817d87a03d2608"
  >
    <svg
      style="overflow: hidden; height: 0px; width: 0px; opacity: 0;"
    >
      <defs>
        <filter
          color-interpolation-filters="sRGB"
          height="300%"
          id="filter-95f111bf6c90801da04c3e2220c21389"
          width="300%"
          x="-100%"
          y="-100%"
        >
          <feflood
            flood-color="rgba(0,0,0,.5)"
            result="flooded-0"
          />
          <femorphology
            in="SourceAlpha"
            operator="dilate"
            radius="3"
            result="dilated-0"
          />
          <feoffset
            dx="0.05"
            dy="0.05"
            in="dilated-0"
            result="offset-0"
          />
          <fegaussianblur
            in="offset-0"
            result="blurred-0"
            stdDeviation="0.0625"
          />
          <fecomposite
            in="flooded-0"
            in2="blurred-0"
            operator="in"
            result="shadow-0"
          />
          <femerge
            result="filter-0"
          >
            <femergenode
              in="shadow-0"
            />
            <femergenode
              in="SourceGraphic"
            />
          </femerge>
          <feflood
            flood-color="rgba(0,0,0,.5)"
            result="flooded-1"
          />
          <femorphology
            in="SourceAlpha"
            operator="dilate"
            radius="0"
            result="dilated-1"
          />
          <feoffset
            dx="0.125"
            dy="0.125"
            in="dilated-1"
            result="offset-1"
          />
          <fegaussianblur
            in="offset-1"
            result="blurred-1"
            stdDeviation="0.25"
          />
          <fecomposite
            in="flooded-1"
            in2="blurred-1"
            operator="in"
            result="shadow-1"
          />
          <femerge
            result="filter-1"
          >
            <femergenode
              in="shadow-1"
            />
            <femergenode
              in="SourceGraphic"
            />
          </femerge>
          <femerge>
            <femergenode
              in="filter-1"
            />
            <femergenode
              in="filter-0"
            />
          </femerge>
        </filter>
      </defs>
    </svg>
  </div>
</body>
`;
